File: php/emit_log_direct.php
<?php

require_once(__DIR__ . '/lib/php-amqplib/amqp.inc');

$connection = new AMQPConnection('localhost', 5672, 'guest', 'guest');
$channel = $connection->channel();


$channel->exchange_declare('direct_logs', 'direct', false, false, false);

$severity = $argv[1];
if(empty($severity)) $severity = "info";

$data = implode(' ', array_slice($argv, 2));
if(empty($data)) $data = "Hello World!";

$msg = new AMQPMessage($data);

$channel->basic_publish($msg, 'direct_logs', $severity);

echo " [x] Sent ",$severity,':',$data," \n";

$channel->close();
$connection->close();

?>

File: rust/src/bin/emit_log_direct.rs
use lapin::{
    options::*, types::FieldTable, BasicProperties, Connection, ConnectionProperties, ExchangeKind,
};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<_> = std::env::args().skip(1).collect();
    let severity = args.first().map(String::as_str).unwrap_or("info");
    let message = match args.len() {
        x if x < 2 => b"Hello, world!".to_vec(),
        _ => args[1..].join(" ").into_bytes(),
    };

    let addr = "amqp://127.0.0.1:5672";
    let conn = Connection::connect(addr, ConnectionProperties::default()).await?;
    let channel = conn.create_channel().await?;

    channel
        .exchange_declare(
            "direct_logs",
            ExchangeKind::Direct,
            ExchangeDeclareOptions::default(),
            FieldTable::default(),
        )
        .await?;

    channel
        .basic_publish(
            "direct_logs",
            severity,
            BasicPublishOptions::default(),
            message.clone(),
            BasicProperties::default(),
        )
        .await?;

    println!(
        " [x] Sent {}:{:?}",
        severity,
        std::str::from_utf8(&message)?
    );

    conn.close(0, "").await?;

    Ok(())
}
